### ISSUE TITLE
numpy.argmin() fails to find the minimum for the most negative integer (migrated from Trac #1878)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3426]
Original ticket http://projects.scipy.org/numpy/ticket/1878
Reported 2011-06-22 by trac user tlatorre, assigned to unknown.

```
#!python
>>> a = np.array([-2**7, -2**7+1, -2**7+2], dtype=np.int8)
>>> a
array([-128, -127, -126], dtype=int8)
>>> np.argmin(a)
1
>>> b = np.array([-2**15, -2**15+1, -2**15+2], dtype=np.int16)
>>> b
array([-32768, -32767, -32766], dtype=int16)
>>> np.argmin(b)
1
>>> c = np.array([-2**31, -2**31+1, -2**31+2], dtype=np.int32)
>>> c
array([-2147483648, -2147483647, -2147483646], dtype=int32)
>>> np.argmin(c)
1
```

**Comment by thouis** [comment]
Attachment in Trac by trac user alfonsonp, 2011-07-22: [0001-ABA-Test-and-fix-for-argmin-in-integers.patch](http://projects.scipy.org/numpy/attachment/ticket/1878/0001-ABA-Test-and-fix-for-argmin-in-integers.patch)

**Comment by thouis** [comment]
Comment in Trac by trac user alfonsonp, 2011-07-22

I have just sent a fix for this issue (patch 0001-ABA-Test-and-fix-for-argmin-in-integers.patch). The fix consist of adding one to array if array is integer. I have also included test cases.

Besides, in argmin function, I changed error message, because I think that error message should be that argmin is not supported, not argmax (although argmax was the real function).

**Comment by thouis** [comment]
Comment in Trac by trac user alfonsonp, 2011-07-22

Also, with python 2.6.6 and numpy 1.4.1, without the patch, numpy.argmin works well for dtype=np.int8, but it does not work for dtype=np.int16 and dtype=np.int32

**Comment by thouis** [comment]
Attachment in Trac by trac user alfonsonp, 2011-07-26: [0001-ABA-Test-and-fix-for-argmin-in-integers-v2.patch](http://projects.scipy.org/numpy/attachment/ticket/1878/0001-ABA-Test-and-fix-for-argmin-in-integers-v2.patch)

**Comment by thouis** [comment]
Comment in Trac by trac user alfonsonp, 2011-07-26

I have produced a new patch that fix this issue (0001-ABA-Test-and-fix-for-argmin-in-integers-v2.patch). The only difference with the previous one is the added code place. In this way, it avoids a possible memory leak for obj variable if the return of PyArray_EnsureAnyArray (for op) is NULL.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-05-02

This has been fixed in 1.7 and in the upcoming 1.6.2 bugfix release. I'll put in the test. Thanks, and sorry to take so long.
